wpiutil/src/main/java/edu/wpi/first/util/sendable2/SendableTable.java

```diff
@@ -0,0 +1,250 @@
+
+  void setDouble(String name, double value);
+
+  void publishDouble(String name, DoubleSupplier supplier);
+
+  DoubleConsumer publishDouble(String name);
+
+  void subscribeDouble(String name, DoubleConsumer consumer);
+
+  void setString(String name, String value);
+
+  void publishString(String name, Supplier<String> supplier);
+
+  Consumer<String> publishString(String name);
+
+  void subscribeString(String name, Consumer<String> consumer);
+
+  void setRaw(String name, String typeString, byte[] value, int start, int len);
+
+  void setRaw(String name, String typeString, ByteBuffer value, int start, int len);
+
+  void publishRaw(String name, String typeString, Supplier<byte[]> supplier);
+
+  void publishRawBB(String name, String typeString, Supplier<ByteBuffer> supplier);
+
+  Consumer<byte[]> publishRaw(String name, String typeString);
+
+  Consumer<ByteBuffer> publishRawBB(String name, String typeString);
+
+  void subscribeRaw(String name, String typeString, Consumer<byte[]> consumer);
+
+  default <T> void setStruct(String name, T value, Struct<T> struct) {
+    addSchema(struct);
+    // this is inefficient; implementations should cache StructBuffer
+    StructBuffer<T> buf = StructBuffer.create(struct);
+    ByteBuffer bb = buf.write(value);
+    setRaw(name, struct.getTypeString(), bb, 0, bb.position());
+  }
+
+  default <T> void publishStruct(String name, Struct<T> struct, Supplier<T> supplier) {
+    addSchema(struct);
+    final StructBuffer<T> buf = StructBuffer.create(struct);
+    publishRawBB(name, struct.getTypeString(), () -> {
+      return buf.write(supplier.get());
+    });
+  }
+
+  default <T> Consumer<T> publishStruct(String name, Struct<T> struct) {
+    addSchema(struct);
+    final StructBuffer<T> buf = StructBuffer.create(struct);
+    final Consumer<ByteBuffer> consumer = publishRawBB(name, struct.getTypeString());
+    return (T value) -> {
+      consumer.accept(buf.write(value));
+    };
+  }
+
+  default <T> void subscribeStruct(String name, Struct<T> struct, Consumer<T> consumer) {
+    addSchema(struct);
+    final StructBuffer<T> buf = StructBuffer.create(struct);
+    subscribeRaw(name, struct.getTypeString(), (data) -> {
+      consumer.accept(buf.read(data));
+    });
+  }
+
+  <T> void setProtobuf(String name, T value, Protobuf<T, ?> proto);
+
+  <T> void publishProtobuf(String name, Protobuf<T, ?> proto, Supplier<T> supplier);
+
+  <T> Consumer<T> publishProtobuf(String name, Protobuf<T, ?> proto);
+
+  <T> void subscribeProtobuf(String name, Protobuf<T, ?> proto, Consumer<T> consumer);
+
+  <T> SendableTable addSendable(String name, T obj, Sendable<T> sendable);
+
+  // SendableTable getSendable(String name);
+
+  SendableTable getChild(String name);
+
+  void setPublishOptions(SendableOption... options);
+
+  void setSubscribeOptions(SendableOption... options);
+
+  /**
+   * Gets the current value of a property (as a JSON string).
+   *
+   * @param name name
```

Comment:
It's very early for this, but this parameter description should be improved and unambiguous that it's the name of the topic to apply the metadata to.

I'm not a fan of splitting metadata from the data, but I admit that unifying would cause an overload explosion. What's the expected behavior if the topic doesn't exist (or if it's a Sendable)? If there is an ordering needed between the data call and the metadata call, it should be documented.

Also, I'm inferring that this won't be doing table "metadata" like the old SmartDashboard dot-prefixed entries? Though an implementation for that would be akin to `setString` etc, so I guess specific functions aren't needed.

Reply:
Yes, we'll probably need to implement caching of it to handle it being set prior to the publish call.  We'll have to do the same thing for publish/subscribe options.  It's more complex on the backend implementation, but makes the API a lot simpler to separate the two, otherwise these functions get very unwieldy to use.

Old-style metadata can still be done just through normal setString et al.